### DISCUSSION TITLE
C# 'AssociatedObject is { }' Statement

### DISCUSSION THREAD

**Discussion by najak3d** [AvaloniaUI/Avalonia#7485] [Q&A] [answered]
This is truly just a C# question, but since I ONLY find it inside an Avalonia Add-on from @wieslawsoltes (CustomBehaviors), and cannot find this anywhere else on the internet, I'll ask it here:

Here's the code in question:
```c#
    protected override void OnDetaching()
    {
        base.OnDetaching();
        if (AssociatedObject is { })
        {
            AssociatedObject.DoubleTapped -= DoubleTapped; 
        }
    }
```

Is the expression "(AssociatedObject is { })" the exact equivalent to "(AssociatedObject != null)"?  If not, then how does it differ in function?

As far as I can tell, this "is {...}" notation is primarily for filtering by Property Values, such as:

   (AssociatedObject is { IsActive:  true }) 
           
 would work the same as:

   (AssociatedObject.IsActive)
  
Our team is just now starting to emerge from old C# standards into the new syntax for C#7 and on.   (Our code base had to compile for WindowsCE, CompactFramework! - which is .NET 3.5.)     And Avalonia is using all the new standards, but for this "is { }" example, I can find no documentation examples of this usage where the "{ }" are empty.

**Comment by wieslawsoltes** [comment, marked as answer]
The `is { }` is from when C# did not have `is not null` syntax. When adding nullability annotation it was really hard to tell if AssociatedObject should be marked as it can be null but actually it's possible when you do use Dispatcher jobs.

**Reply by najak3d** [reply]
Perfect.... so it's equivalent to "is not null".

As of now, I'm not sure how "is not null" differs from "!= null".     Are those the same too?

**Reply by wieslawsoltes** [reply]
> Perfect.... so it's equivalent to "is not null".
> 
> As of now, I'm not sure how "is not null" differs from "!= null". Are those the same too?


`When you match an expression against null, the compiler guarantees that no user-overloaded == or != operator is invoked.`

https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/is

**Reply by maxkatz6** [reply]
Yes.
Although "!= null" will use overridden Equals method, if it's available. While "is not null" will always test reference. Same for " is {}". Very minor benefit though.

**Comment by robloo** [comment]
I kind-of agree this is not the greatest syntax and wish it wasn't used in Avalonia. It was always better to use something like:
 * `!object.ReferenceEquals(AssociatedObject, null)` which is kind-of old or
 * `!(AssociatedObject is null)` which is newer

Note that as @maxkatz6 said, `!= null` will use the overridden Equals method if it exists. That can get you into trouble when you are doing null checks within the overridden Equals method itself so one of the other techniques mentioned must be used in that case.
